notebooks/21_reinforcement_learning/index.md

```diff
@@ -0,0 +1,522 @@
+$\frac{( -1  -1 + 10) + ( -1 -1 + 10 ) }{ 2} = +8$
+<br />
+
+**state C:** &nbsp; You can see C in all episodes $\rightarrow$
+$\frac{( -1 + 10 )+( -1+10 ) + ( -1 + 10 )+( -1 -10 )} {4} = +4$
+
+**state D:** &nbsp; You can see D in The first three states $\rightarrow$ &nbsp;
+$\frac{(+10+10+10)}{3} =+10$
+
+If you pay more attention, you will notice that the problem with this method is, we can move to C both from E and B, but we find very different values for them. The reason for this difference is the lack of sample. we also express the values of each state independent from other states.
+<br /><br />
+
+<div id='AdvantagesofDirectUtilityEstimation'/>
+
+## Advantages of Direct Utility Estimation
+
+* Easy to understand
+
+* Does not require any knowledge of T, R 
+
+* Computes the correct average values, using just sample transitions
+
+
+<br />
+
+<div id='DisadvantagesofDirectUtilityEstimation'/>
+
+## Disadvantages of Direct Utility Estimation
+* Wastes information about state connections 
+
+* Each state must be learned separately 
+
+* Takes a long time to learn
+
+* Converges very slowly
+
+&nbsp;&nbsp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;Why?
+
+&nbsp;&nbsp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;  **-** Does not exploit the fact that utilities of states are not independent 
+
+&nbsp;&nbsp;&ensp;&ensp;&ensp;&ensp;&ensp;&ensp;  **-** Utilities follow the Bellman equation
+
+&ensp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; $V_{\pi} (s) =R(S)+\gamma \sum_{\acute{s}} T(s,\pi(s),{\acute{s}})V_{\pi}({\acute{s}})$
+
+
+
+
+
+**Question**: How to approximate E[X] (The right side of the equation below) based on model-free learning?
+
+$V^\pi_{k+1}(s) \leftarrow \sum_{{\acute{s}}} T(s,\pi(s),\acute{s} )[R(s,\pi(s),\acute{s} )+\gamma V^\pi_{k} ]$
+
+We cannot use this equation directly since we do not have T and R, but we can use the average of the returned value ($[R(s,\pi(s),\acute{s} )+\gamma V^\pi_{k} ]$). 
+So we can change our formula to the following form.
+
+$V^\pi (s) \approx \frac{1}{N} \sum [R(s,\pi(s),\acute{s})+\gamma  V^\pi(\acute{s})]$
+
+If you pay a little attention, you will notice that there is a problem, so what is the problem with this approach? In one state, an action cannot be repeated N times, because you move on to another state as soon as an action is taken and it will be physically impossible be able to perform another action in the same state. Now, what is the solution?
+
+You can take the average in more than one shot, for example, take different actions in multiple episodes and update the average as you arrive to the desired state.
+For example, if the average is $\varphi$ up to this point, the updated average would be:
+
+$\varphi_{new} = \varphi n + \frac {x} {n+1}$
+
+In which $n$ is the number of data you have collected.
+
+
+
+
+
+
+<div id='TemporalDifference(TD)Learning'/>
+
+# Temporal Difference (TD) Learning

```

Comment:
Provide the reader with some preparation & introduction before jumping to a topic. Work a bit more on the structure of your notes

Reply:
Should we add introduction before each subtopic? We have already divided the main topic into subtopics and there is an introduction at the beginning of the lecture note . There is also a definition at the beginning of some subtopics.